modules/junit-jupiter/src/main/java/org/testcontainers/junit/jupiter/TestcontainersExtension.java

```diff
@@ -0,0 +1,125 @@
+package org.testcontainers.junit.jupiter;
+
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ExtensionContext.Store.CloseableResource;
+import org.junit.jupiter.api.extension.TestInstancePostProcessor;
+import org.junit.platform.commons.support.AnnotationSupport;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ReflectionUtils;
+import org.testcontainers.lifecycle.Startable;
+
+import java.lang.reflect.Field;
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+class TestcontainersExtension implements TestInstancePostProcessor, BeforeEachCallback {
+
+    private static final Namespace NAMESPACE = Namespace.create(TestcontainersExtension.class);
+
+    private static final String TEST_INSTANCE = "testInstance";
+
+    @Override
+    public void postProcessTestInstance(final Object testInstance, final ExtensionContext context) {
+        ExtensionContext.Store store = context.getStore(NAMESPACE);
+        store.put(TEST_INSTANCE, testInstance);
```

Comment:
This should be a singleton, right? Meaning a `store.get(TEST_INSTANCE)` should always return the same instance?

Reply:
@marcphilipp / @sbrannen What's your stance on this in case of parallel execution?

Reply:
The `ExtensionContext` passed to `TestInstancePostProcessors` is the one of the test class. Thus, unless I'm missing something, this should not work with parallel test execution.

Reply:
What is purpose of storing the test instance in the first place?

Reply:
It is later used to access outer instances of `@Nested` test cases.

Reply:
Does the parent instance know its nested test cases?
Else what about saving the parent `testInstance` under canonicalName as key and then navigate the canonicalName of the inner class up? Very hacky, but could work until we get the API.

Reply:
There is a hack in JUnit `ReflectionUtils` that relies on some JDK internals: https://github.com/junit-team/junit5/blob/master/junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java#L700-L717

Since it is private, we could copy it to the Testcontainers extension until a supported API has been implemented in JUnit.

Reply:
What about 
```
store.put(testInstance.getClass().getCanonicalName(), testInstance);
```

and 
```
String innerClassName = getClass().getCanonicalName();
String outerClassName = innerClassName.substring(0, innerClassName.lastIndexOf("."));
ctx.getStore(NAMESPACE).remove(outerClassName)
```
?

Not a working example, just as an idea for discussion.

Reply:
Only if parallel execution happens inside the nested class. 
I was assuming, parallel execution will be on outer class level.

Reply:
I think we need to wait for https://github.com/junit-team/junit5/issues/1618

But I also want to dig into the junit code to be sure this does really not work in parallel mode.

Reply:
We can also merge and add a disclaimer, that parallel execution isn't supported ATM. Better than leaving this PR blocked for a longer time.